File: src/amazon/leetcode438_FindAllAnagrams.py


class Solution(object):
    """
    @ Amazon

    Hash Table

    Given a string s and a non-empty string p, find all the start indices of p's anagrams in s.

    Strings consists of lowercase English letters only and the length of both strings s and p
    will not be larger than 20,100.

    The order of output does not matter.

    Example 1:

    Input:
    s: "cbaebabacd" p: "abc"

    Output:
    [0, 6]

    Explanation:
    The substring with start index = 0 is "cba", which is an anagram of "abc".
    The substring with start index = 6 is "bac", which is an anagram of "abc".
    """

    def findAnagrams_ASCII(self, s, p): # O(n) time, O(1) space
        """
        :type s: str
        :type p: str
        :rtype: List[int]
        """
        res = []
        m, n = len(s), len(p)
        if m < n: return res

        # hash table of string p
        table = [0] * 128
        # record each character in p to hash table
        for c in p:
            table[ord(c)] += 1

        # sliding window
        left = right = 0
        count = len(p)
        while right < m:
            key = ord(s[right])
            # current hash value >= 1 means the character is existing in p
            if table[key] >= 1:
                count -= 1
            table[key] -= 1
            right += 1

            # when b is down to 0, we found the right anagram,
            # then save left
            if count == 0:
                res.append(left)

            if right - left == len(p):
                key = ord(s[left])
                # only increase n if the character is in p
                if table[key] >= 0:
                    count += 1
                left += 1
                table[key] += 1

        return res

    def findAnagrams_HashTable(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: List[int]
        """
        res = []
        if not s or not p or len(s) == 0 or len(p) == 0 or len(s) < len(p):
            return res

        m, n = len(s), len(p)
        # create a hash table
        table = {}
        for c in p:
            if c in table:
                table[c] += 1
            else:
                table[c] = 1

        left = right = count = 0
        while (right < m):

            # check the char at right
            c = s[right]
            if c in table:
                # update value related to key
                table[c] -= 1
                if table[c] == 0:
                    # match a char in p
                    count += 1

            # check the window size
            if (right >= n):
                c = s[left]
                if c in table:
                    table[c] += 1
                    if table[c] == 1:
                        count -= 1
                left += 1

            # output left
            if count == len(table):
                res.append(left)

            right += 1

        return res


if __name__ == "__main__":
    s = "cbaebabacd"
    p = "abc"
    res = Solution().findAnagrams_slidingWindow(s, p)
    print res

File: src/amazon/java_version/leetcode438_FindAllAnagrams.java


public class Solution {
    public List<Integer> findAnagrams_ASCIITable(String s, String p) {

        List<Integer> res = new ArrayList<>();

        if (s == null || s.length() == 0 || p == null || p.length() == 0 || s.length() < p.length()) {
            return res;
        }

        // create a ASCII table
        int[] table = new int[256];
        for (char c : p.toCharArray()) {
            table[c] += 1;
        }

        int left = 0, right = 0;
        int count = p.length();
        while (right < s.length()) {

            // check right
            if (table[s.charAt(right)] >= 1) {
                count -= 1;
            }
            // update table
            table[s.charAt(right)] -= 1;
            // move right
            right += 1;

            if (count == 0) {
                res.add(left);
            }

            // check and move left
            if (right - left == p.length()) {
                // recover the number
                if (table[s.charAt(left)] >= 0) {
                    count += 1;
                }
                // recover table
                table[s.charAt(left)] += 1;
                // move left
                left += 1;
            }
        }

        return res;
    }

    public List<Integer> findAnagrams_HashTable(String s, String p) {

        List<Integer> res = new ArrayList<>();

        if (s == null || s.length() == 0 || p == null || p.length() == 0 || s.length() < p.length()) {
            return res;
        }

        // create a hash table
        Map<Character, Integer> map = new HashMap<>();
        for (char c : p.toCharArray()) {
            if (map.containsKey(c)) {
                map.put(c, map.get(c)+1);
            } else {
                map.put(c, 1);
            }
        }

        int match = 0;
        for (int i=0; i<s.length(); i++) {
            char c = s.charAt(i);
            if (map.containsKey(c)) {
                map.put(c, map.get(c)-1);
                if (map.get(c) == 0) {
                    match += 1;
                }
            }

            if (i >= p.length()) {
                c = s.charAt(i - p.length());
                if (map.containsKey(c)) {
                    map.put(c, map.get(c)+1);
                    if (map.get(c) == 1) {
                        match -= 1;
                    }
                }
            }

            if (match == map.size()) {
                res.add(i - p.length() + 1);
            }
        }

        return res;
    }
}